Configuring Read the Docs to publish the MPS manual

rptb1 opened the pull request:
See
- https://github.com/Ravenbrook/mps/issues/98#issuecomment-1379115620
- #118

rptb1 commented:
To do: #125 must consider how to build, publish, and link to the relevant version of the manual on Read the Docs.

rptb1 commented:
Bootstrap issue: If the Read the Docs manual goes away, and someone has a bare MPS tree, how can they read the manual?  The readme.txt does not lead to instructions for how to build and read it.

rptb1 commented:
https://memory-pool-system.readthedocs.io/en/latest/design/index.html is blank, so Read the Docs didn't run Sphinx in the same way as "make html" would.

rptb1 commented:
> https://memory-pool-system.readthedocs.io/en/latest/design/index.html is blank, so Read the Docs didn't run Sphinx in the same way as "make html" would.

Examining the Read the Docs build output at https://readthedocs.org/projects/memory-pool-system/builds/19364054/ reveals that it's invoking Sphinx with
```
python -m sphinx -T -E -b html -d _build/doctrees -D language=en . ../../_readthedocs/html
```
whereas we invoke it with https://github.com/Ravenbrook/mps/blob/4108cc4257c6961367e410da398e4692d5cce7ce/manual/Makefile#L48

So Read the Docs is building with "manual/source" as the current working directory.  That causes this glob to find no design documents https://github.com/Ravenbrook/mps/blob/4108cc4257c6961367e410da398e4692d5cce7ce/manual/source/extensions/mps/designs.py#L162

rptb1 commented:
> That causes this glob to find no design documents

Fixed in https://github.com/Ravenbrook/mps/pull/141/commits/409aea106e914745b75e544629dbac03e42be578 . You can see that the designs are converted in https://readthedocs.org/projects/memory-pool-system/builds/19364368/ output from the Sphinx command and at https://memory-pool-system.readthedocs.io/en/branch-2023-02-02-readthedocs/design/index.html

rptb1 commented:
> Bootstrap issue: If the Read the Docs manual goes away, and someone has a bare MPS tree, how can they read the manual? The readme.txt does not lead to instructions for how to build and read it.

Fixed in https://github.com/Ravenbrook/mps/pull/141/commits/380b43dee2369e98c6b59fb66bbab07ae2b240bd which also closes #144

rptb1 commented:
Review suggestion: Check https://memory-pool-system.readthedocs.io/en/branch-2023-02-02-readthedocs/index.html against https://www.ravenbrook.com/project/mps/master/manual/html/ to make sure building with Read the Docs hasn't broken anything.  They ought to be identical in content, except for changes made in this branch.

rptb1 commented:
@UNAA008 spotted a couple of [warnings from Read the Docs](https://readthedocs.org/projects/memory-pool-system/builds/19364651/):
```
WARNING: while setting up extension extensions.mps: directive 'deprecated' is already registered, it will be overridden
WARNING: while setting up extension extensions.mps: node class 'note' is already registered, its visitors will be overridden
```

Investigate.

rptb1 commented:
> @UNAA008 spotted a couple of [warnings from Read the Docs](https://readthedocs.org/projects/memory-pool-system/builds/19364651/)

https://github.com/Ravenbrook/mps/blob/631d30ea5cc1307f801ec2b6d0c3444e6f411172/manual/source/extensions/mps/__init__.py is not well commented, and it's hard to discern the intention, but these are definitely deliberate overrides.

I think adding commentary or working out whether and how to suppress these warnings is outside the scope of this pull request and not very important.

rptb1 commented:
The manual on Read the Docs doesn't seem to have a search.  Investigate.

rptb1 commented:
GitHub deleted my express review record.  Reconstructing...

We executed [proc.review.express](https://github.com/Ravenbrook/mps/blob/936f0585c61031e3b19436fd889721f6ab048501/procedure/review.rst#12-express-review).

1. @UNAA008 and @thejayps attending.
2. @UNAA008 had a PI: "TODOs" are for nice-level stuff.
3. @rptb1 has a new PI: Press the Comment button to save the review record before doing anything else with your tab.
4. We found no major defects and three minor defects.
5. @rptb1 I wonder if we could have a general duplicate text finder to prevent https://github.com/Ravenbrook/mps/pull/141#discussion_r1108590540
6. @rptb1 Iwbni we had a global tag checker to prevent https://github.com/Ravenbrook/mps/pull/141#discussion_r1108592239
7.

rptb1 commented on manual/build.txt:
```diff
@@ -94,6 +94,30 @@ If you want to do anything beyond these simple cases, use the MPS build
+manual, like this::
+
+  cd manual
+  make html
```
Fixed in 170d805961ffe129d3456df6f3724b65c1d5bacf

rptb1 commented on readme.txt:
```diff
@@ -133,6 +137,7 @@ Document History
 2014-07-04  GDR_   Link to hotfix for WOW64 bug.
 2016-03-24  RB_    Adding support for FreeBSD with Clang/LLVM.
 2020-05-22  PNJ_   Changed to BSD 2-clause licence.
+2023-02-02  RB_    Migrating from Ravenbrook Perforce to GitHub.
```
Fixed in 170d805961ffe129d3456df6f3724b65c1d5bacf

rptb1 commented on manual/source/extensions/mps/designs.py:
```diff
@@ -43,22 +46,43 @@
+# design.mps.doc.fmt.ref
 mps_ref = re.compile(r'`(\.[a-z][A-Za-z.0-9_-]*[A-Za-z0-9])`_(?:        )?')
+
+# design.mps.fmt.function-decl
```
Fixed in 170d805961ffe129d3456df6f3724b65c1d5bacf

thejayps reviewed (approved):
proc.review.exit.pass

https://github.com/Ravenbrook/mps/pull/141#issuecomment-1433253292

rptb1 commented:
Executing [proc.merge.pull-request](https://github.com/Ravenbrook/mps/blob/8a88acdc976c479cf55fdd6ef0935f0d55eef910/procedure/pull-request-merge.rst).

1. Start time 16:17.
2. An automated test case isn't feasible because it would need to publish a readme.txt via GitHub and check that the link to the manual was broken.
3. End time 16:22.
4. Merging procedure took 5 minutes.

rptb1 commented:
> * @UNAA008 had a PI: "TODOs" are for nice-level stuff.

Added to https://github.com/Ravenbrook/mps/issues/140#issuecomment-1434317186

> * @rptb1 has a new PI: Press the Comment button to save the review record before doing anything else with your tab.

Fixed in 6fa80f4e1dee1e6742cb829ef5d8d22101fb30d9

> * @rptb1 I wonder if we could have a general duplicate text finder

Created #160 

> * @rptb1 Iwbni we had a global tag checker to prevent ..

Created #161
